### DISCUSSION TITLE
Customizing String formats

### DISCUSSION THREAD

**Discussion by dbire** [schemathesis/schemathesis#1943] [Q&A]
Hello,

is there a way too restrict ALL strings used data generation from being " " or "\r" any of these kind of characters which are causing 400s ?
At the end I am adding more and more rules in my hook to use my own rule on each element of the body, which is not really satisfying.
What would be the best practice ?

**Comment by Stranger6667** [comment]
Hi! Sorry for the late answer.

In the recent versions, I've implemented some config options specifically for this. Here are the [docs](https://schemathesis.readthedocs.io/en/stable/data-generation.html#generating-strings). It is not that flexible though, but allows for the exclusion of some characters (like `\x00` and non-ASCII characters). It is possible to use any Python codec described [here](https://docs.python.org/3/library/codecs.html).

It also should be possible to create a custom codec (as described in Python docs). My guess is that it would be quite cumbersome for the sake of exclusion of certain characters, but still possible. So, maybe Schemathesis can ship some shortcut for it
